app/models/concerns/unparsed_run.rb

```diff
@@ -13,154 +18,286 @@ def parse_into_db
+          )
+          segment.send(
+            "#{timing}time_gold=",
+            segment.send("#{timing}time_duration_ms") <= segment.send("#{timing}time_shortest_duration_ms")
+          )
+          segment.send(
+            "#{timing}time_reduced=",
+            false
+          )
+        end
+
+        container[lsc_segment] = segment
+        previous_segment = segment
+        next if container.size < MAX_CONTAINER_SIZE
+
+        result = write_segments(container.values)
+        process_segment_histories(container.keys, result.ids)
+        container = {}
       end
 
-      segments.import(
-        %i[
-          run_id
-          segment_number
-          name
-
-          realtime_start_ms
-          realtime_end_ms
-          realtime_duration_ms
-          realtime_shortest_duration_ms
-          realtime_skipped
-          realtime_reduced
-          realtime_gold
-
-          gametime_start_ms
-          gametime_end_ms
-          gametime_duration_ms
-          gametime_shortest_duration_ms
-          gametime_skipped
-          gametime_reduced
-          gametime_gold
-        ],
-        segs
-      )
-
-      segs
+      return if container.blank?
+
+      result = write_segments(container.values)
+      process_segment_histories(container.keys, result.ids)
     end
 
-    def write_segment_histories(segs)
-      ids = segments.order(:segment_number).pluck(:id)
-      histories = []
+    def write_segments(segments)
+      segments.each do |seg|
+        %w[real game].each do |timing|
+          %w[time_end_ms time_duration_ms time_shortest_duration_ms].each do |field|
+            value = seg.send("#{timing}#{field}")
+            seg.send("#{timing}#{field}=", zero_to_nil(value))
+          end
+        end
+      end
 
-      segs.each.with_index do |seg, i|
-        return nil if seg[:history].blank?
+      result = Segment.import(segments)
+      return result unless result.failed_instances.any?
 
-        seg[:history].each do |history|
-          next unless history[:attempt_number].to_i.positive?
+      raise UnparsableRun
+    end
 
-          histories << SegmentHistory.new(
-            segment_id:           ids[i],
-            attempt_number:       history[:attempt_number].to_i,
-            realtime_duration_ms: history[:realtime_duration_ms],
-            gametime_duration_ms: history[:gametime_duration_ms]
-          )
+    def process_segment_histories(lsc_segments, segment_ids)
+      container = []
+
+      lsc_segments.each_with_index do |seg, i|
+        seg.segment_history.iter.with do |iter|
+          until (history_element = iter.next).nil?
+            next unless history_element.index.positive?
+
+            container << SegmentHistory.new(
+              segment_id:           segment_ids[i],
+              attempt_number:       history_element.index,
+              realtime_duration_ms: history_element.time.real_time&.total_seconds&.*(1000) || 0,
```

Comment:
I wasn't sure if these two needed to be `zero_to_nil`'d or not.

Reply:
I'm not sure if they _need_ to be, but I do think they should be just in case -- I don't see any situation where a segment was literally completed in zero time, but I do see situations where (a) livesplit-core reports zero times for absent segments or (b) it doesn't currently but will in the future, so afaict there's no harm in `zero_to_nil`ing it.